commit e9449d85c67127d6f9d01aad8963d567ab02cb96
Author: Divy Le Ray <divy@chelsio.com>
Date:   Sat Feb 13 09:44:35 2010 +0000

    cxgb3: fix link flap
    
    The driver is expected to report that the link is up
    when the phy Rx signal is established and the mac
    has not detected a link fault.
    The code is however broken, the driver does not check the link fault
    status when the phy link status changes.
    The link fault status being checked within a short period of time,
    it leads to link up/link down events.
    
    Signed-off-by: Divy Le Ray <divy@chelsio.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/drivers/net/cxgb3/t3_hw.c b/drivers/net/cxgb3/t3_hw.c
index 032cfe0..3ab9f51 100644
--- a/drivers/net/cxgb3/t3_hw.c
+++ b/drivers/net/cxgb3/t3_hw.c
@@ -1262,7 +1262,8 @@ void t3_link_changed(struct adapter *adapter, int port_id)
 		lc->fc = fc;
 	}
 
-	t3_os_link_changed(adapter, port_id, link_ok, speed, duplex, fc);
+	t3_os_link_changed(adapter, port_id, link_ok && !pi->link_fault,
+			   speed, duplex, fc);
 }
 
 void t3_link_fault(struct adapter *adapter, int port_id)
